c7n/resources/iam.py

```diff
@@ -94,6 +94,33 @@ def _user_keys(resource):
                     break
         return matched
 
+# Mfa-device filter for iam-users
+@User.filter_registry.register('mfa-device')
+class UserMfaDevice(ValueFilter):
+
+    schema = type_schema('mfa-device', rinherit=ValueFilter.schema)
+
+    def process(self, resources, event=None):
+
+        def _user_mfa_devices(resource):
+            client = local_session(self.manager.session_factory).client('iam')
+            resource['MfaDevices'] = client.list_mfa_devices(
+                UserName=resource['UserName'])['MFADevices']
+
+        with self.executor_factory(max_workers=2) as w:
+            query_resources = [
+                r for r in resources if 'MfaDevices' not in r]
+            self.log.debug("Querying %d users' mfa devices" % len(query_resources))
+            list(w.map(_user_mfa_devices, query_resources))
+
+        matched = []
+        for r in resources:
+            for p in r['MfaDevices']:
```

Comment:
possibly want to match on full set of mfadevices.. ala self.match(r['MfaDevices']) so that you can filter to users without mfa devices.